tchap_red_list/__init__.py

```diff
@@ -11,36 +11,211 @@
+        desired_status = content.get("hide_profile") or False
+        current_status, _ = await self._get_user_status(user_id)
+
+        if current_status == desired_status:
+            return
+
+        # Add or remove the user depending on whether they want their profile hidden.
+        if desired_status is True:
+            await self._add_to_red_list(user_id)
+        else:
+            await self._remove_from_red_list(user_id)
+
+    async def _maybe_change_membership_in_discovery_room(
+        self, user_id: str, membership: str
+    ) -> None:
+        """Change a user's membership in the discovery room.
+
+        Does nothing if no discover room has been configured.
+
+        Args:
+            user_id: the user to change the membership of.
+            membership: the membership to set for this user.
+        """
+        if self._config.discovery_room is None:
+            return
+
+        await self._api.update_room_membership(
+            sender=user_id,
+            target=user_id,
+            room_id=self._config.discovery_room,
+            new_membership=membership,
+        )
+
+    async def check_user_in_red_list(self, user_profile: UserProfile) -> bool:
+        """Check if a user should be in the red list, which means they need to be hidden
+        from local user directory search results.
+        Implements the check_username_for_spam spam checker callback.
+        """
+        user_in_red_list, _ = await self._get_user_status(user_profile["user_id"])
+        return user_in_red_list
+
+    async def _setup_db(self) -> None:
+        """Create the table needed to store the red list data."""
+
+        def setup_db_txn(txn: LoggingTransaction) -> None:
+            sql = """
+            CREATE TABLE IF NOT EXISTS tchap_red_list(
+                user_id TEXT PRIMARY KEY,
+                because_expired BOOLEAN NOT NULL DEFAULT FALSE
+            );
+            """
+            txn.execute(sql, ())
+
+        await self._api.run_db_interaction(
+            "tchap_red_list_setup_db",
+            setup_db_txn,
+        )
+
+    async def _add_to_red_list(
+        self,
+        user_id: str,
+        because_expired: bool = False,
+    ) -> None:
+        """Add the given user to the red list.
+
+        Args:
+            user_id: the user to add to the red list.
+            because_expired: whether the user is being added as a result of a change in
+                their account validity status.
+        """
+
+        def _add_to_red_list_txn(txn: LoggingTransaction) -> None:
+            sql = """
+            INSERT INTO tchap_red_list(user_id, because_expired) VALUES (?, ?)
+            """
+            txn.execute(sql, (user_id, because_expired))
+
+            self._get_user_status.invalidate((user_id,))
+
+        await self._api.run_db_interaction(
+            "tchap_red_list_add",
+            _add_to_red_list_txn,
+        )
```

Comment:
We can raise an exception if the user is already in the table. This can happen if there are two concurrent calls to `update_red_list_status` and the user's status changes after the first call to `_get_user_status()`.

Not sure if we want to handle it somehow.

Reply:
It will only happen if there are two concurrent calls for the same user, which I think is highly unlikely to happen in a real life scenario.